File: java/cucumber-archetype/src/main/java/postoffice/services/LabelService.java
package postoffice.services;

import postoffice.models.Label;
import postoffice.models.Package;

import java.util.UUID;

public class LabelService {

    // Method to validate package dimensions and weight
    public boolean validatePackage(Package pkg) {
        if (pkg.getHeight() < 10 || pkg.getWidth() < 10 || pkg.getLength() < 5) {
            throw new IllegalArgumentException("Package is below minimum allowable size");
        }

        if (pkg.getHeight() > 50 || pkg.getWidth() > 50 || pkg.getLength() > 50) {
            throw new IllegalArgumentException("Package is above maximum allowable size");
        }

        if (pkg.getWeight() < 0.01 || pkg.getWeight() > 30) {
            throw new IllegalArgumentException("Package is out of allowable weight range");
        }

        return true;
    }

    // Method to calculate shipping cost
    public double calculateShippingCost(Package pkg) {
        double baseCost = 0;

        // Determine base cost based on delivery type
        switch (pkg.getDeliveryType()) {
            case STANDARD:
                baseCost = 60.0;
                break;
            case EXPRESS:
                baseCost = 120.0;
                break;
            case ECONOMY:
                baseCost = 30.0;
                break;
            default:
                throw new IllegalArgumentException("Invalid delivery type.");
        }

        // Add weight-based fees
        if (pkg.getWeight() > 5.0 && pkg.getWeight() <= 15.0) {
            baseCost += 15.0; // Medium weight fee
        } else if (pkg.getWeight() > 15.0 && pkg.getWeight() <= 30.0) {
            baseCost += 30.0; // Heavy weight fee
        }

        return baseCost;
    }

    // Method to create a label
    public Label createLabel(Package pkg) {
        try {
            validatePackage(pkg); // Ensure validation is done before proceeding
        } catch (IllegalArgumentException e) {
            // Rethrow with the specific message so it's easier for the test to match
            throw new IllegalArgumentException(e.getMessage());
        }

        // Calculate the shipping cost
        double shippingCost = calculateShippingCost(pkg);

        // Generate random UUID
        UUID id = UUID.randomUUID();

        // Create the label
        Label label = new Label(id, pkg, shippingCost);
        label.setTotalCost(shippingCost);

        return label;
    }
}

File: go/PostOffice/services/label_service.go
package services

import (
	"PostOffice/models"
	"errors"
	"fmt"

	"github.com/google/uuid"
)

func ValidatePackage(pkg models.Package) (bool, error) {
	// Check for minimum size
	if pkg.Height < 10 || pkg.Width < 10 || pkg.Length < 5 {
		return false, errors.New("package is below minimum allowable size")
	}

	// Check for maximum size
	if pkg.Height > 50 || pkg.Width > 50 || pkg.Length > 50 {
		return false, errors.New("package is above maximum allowable size")

	}

	if pkg.Weight < 0.01 || pkg.Weight > 30 {
		return false, errors.New("package is out of allowable weight range")
	}

	return true, nil
}

func CalculateShippingCost(pkg models.Package) float32 {
	baseCost := float32(0)

	// Normalize delivery type to handle case-insensitivity
	switch pkg.DeliveryType {
	case "STANDARD":
		baseCost = 60
	case "EXPRESS":
		baseCost = 120
	case "ECONOMY":
		baseCost = 30
	}

	if pkg.Weight > 5 && pkg.Weight <= 15 {
		baseCost += 15
	} else if pkg.Weight > 15 && pkg.Weight <= 30 {
		baseCost += 30
	}

	return baseCost
}

func CreateLabel(pkg models.Package) (models.Label, error) {

	isValid, err := ValidatePackage(pkg)

	if !isValid {
		fmt.Printf("Error validating package: %v", err)
		return models.Label{}, err
	}

	costShipping := CalculateShippingCost(pkg)

	return models.Label{
		ID:        uuid.NewString(),
		Package:   pkg,
		TotalCost: costShipping,
	}, nil
}
